dbcrossbar/src/clouds/gcloud/client.rs

```diff
@@ -222,6 +222,68 @@ impl Client {
         }
     }
 
+    /// Make an HTTP POST request and return the response.
+    async fn post_helper<Body>(
+        &self,
+        url: &Url,
+        body: &Body,
+    ) -> Result<reqwest::Response, ClientError>
+    where
+        Body: fmt::Debug + Serialize + std::marker::Sync,
+    {
+        trace!("POST {}", url);
+        let token = self.token().await?;
+        let wait_options = WaitOptions::default()
+            .backoff_type(BackoffType::Exponential)
+            .retry_interval(Duration::from_secs(10))
+            // Don't retry too much because we're probably classifying some
+            // permanent errors as temporary.
+            .allowed_errors(3);
+        wait(&wait_options, move || {
+            let token = token.clone();
+            async move {
+                let resp_result = self
+                    .client
+                    .post(url.as_str())
+                    .bearer_auth(token.as_str())
+                    .json(body)
+                    .send()
+                    .await;
+                match resp_result {
+                    // The HTTP request failed outright, because of something
+                    // like a DNS error or whatever.
+                    Err(err) => {
+                        // Request and timeout errors look like the kind of
+                        // things we should probably retry. But this is based on
+                        // guesswork not experience.
+                        // In general, it is not safe to retry POST requests, however for BigQuery 503 errors, we can retry.
+                        let temporary = err.is_status()
+                            && err.status() == Some(StatusCode::SERVICE_UNAVAILABLE);
+                        let err: Error = err.into();
+                        let err: ClientError =
+                            err.context(format!("could not POST {}", url)).into();
+                        if temporary {
+                            WaitStatus::FailedTemporarily(err)
+                        } else {
+                            WaitStatus::FailedPermanently(err)
+                        }
+                    }
+                    // We talked to the server and it returned a server-side
+                    // error (50-599). There's a chance that things might work
+                    // next time, we hope.
+                    Ok(resp) if resp.status().is_server_error() => {
+                        WaitStatus::FailedTemporarily(
+                            self.handle_error("POST", url, resp).await,
+                        )
+                    }
```

Comment:
As a general rule, we should not retry things unless we've actually seen them in the logs, and we expect them to be transient. We're seen 503s. The statuses 429, 502 and 504 are also common candidates for retries.

Other 500 errors, especially the generic 500, likely represent an exciting new problem we haven't seen before, and failing outright is probably the best policy.

Reply:
updated to allow (503 | 429 | 502 | 504)